### PR TITLE
Make sure that shared memory object name meets the length requirements

### LINKED ISSUE #63240: MemoryMappedFile.CreateNew on macOS throws System.IO.PathTooLongException
### Description

`MemoryMappedFile.CreateNew` throws a
`System.IO.PathTooLongException: The specified file name or path is too long, or a component of the specified path is too long.`
exception when passing in `null` as the map name (`null` mapName is required on
unix as noted in https://github.com/dotnet/runtime/issues/21863).

Program.cs:

```cs
using System.IO.MemoryMappedFiles;

// Passing null as mapNames are not supported for mmf on unix
using var mmf = MemoryMappedFile.CreateNew(null, 1);
```

When run, will throw:

```
dotnet run --project MemoryMapIssue
Unhandled exception. System.IO.PathTooLongException: The specified file name or path is too long, or a component of the specified path is too long.
   at System.IO.MemoryMappedFiles.MemoryMappedFile.CreateSharedBackingObjectUsingMemory(MemoryMappedProtections protections, Int64 capacity, HandleInheritability inheritability)
   at System.IO.MemoryMappedFiles.MemoryMappedFile.CreateSharedBackingObject(MemoryMappedProtections protections, Int64 capacity, HandleInheritability inheritability)
   at System.IO.MemoryMappedFiles.MemoryMappedFile.CreateCore(FileStream fileStream, String mapName, HandleInheritability inheritability, MemoryMappedFileAccess access, MemoryMappedFileOptions options, Int64 capacity)
   at System.IO.MemoryMappedFiles.MemoryMappedFile.CreateNew(String mapName, Int64 capacity, MemoryMappedFileAccess access, MemoryMappedFileOptions options, HandleInheritability inheritability)
   at System.IO.MemoryMappedFiles.MemoryMappedFile.CreateNew(String mapName, Int64 capacity)
   at Program.<Main>$(String[] args) in /Users/aida/workspace/MemoryMapIssue/Program.cs:line 5
```

### Configuration

- OS: macOS 12.1
- Arch: ARM64

```
dotnet --version
6.0.101
```

```
dotnet --list-runtimes
Microsoft.AspNetCore.App 6.0.1 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
Microsoft.NETCore.App 6.0.1 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
```

```
dotnet --list-sdks
6.0.101 [/usr/local/share/dotnet/sdk]
```

### Regression?

Unknown

### Other information

Following guidance from https://github.com/dotnet/runtime/issues/21863 which
says unix systems do not support named mmf; so passing in `null` as name.

### PR DESCRIPTION
**Pull request by adamsitnik**
~I've not compiled this code yet and I've also no access to macOS ARM64 machine, so I am just opening this PR hoping the CI is going to give me the answer.~

Fixes #63240

### CODE REVIEW

File: src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs [hunk 1]
```diff
@@ -173,8 +173,15 @@ private static FileStream CreateSharedBackingObject(Interop.Sys.MemoryMappedProt
...
         {
-            // The POSIX shared memory object name must begin with '/'.  After that we just want something short and unique.
-            string mapName = string.Create(null, stackalloc char[128], $"/corefx_map_{Guid.NewGuid():N}");
+            // The POSIX shared memory object name must begin with '/'.  After that we just want something short (32) and unique.
+            string mapName = Guid.NewGuid().ToString("N");
+            unsafe
+            {
+                fixed (char* mutable = mapName)
+                {
+                    mutable[0] = '/';
```
Comment (jkotas) [inline src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs:182]: Mutating Strings is not acceptable outside CoreLib.

Comment (adamsitnik) [reply src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs]: @jkotas thanks for the feedback, I've removed the mutation

File: src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs [hunk 2]
```diff
@@ -170,11 +171,17 @@ private static FileStream CreateSharedBackingObject(Interop.Sys.MemoryMappedProt
...
-            // The POSIX shared memory object name must begin with '/'.  After that we just want something short and unique.
-            string mapName = string.Create(null, stackalloc char[128], $"/corefx_map_{Guid.NewGuid():N}");
+            const int MaxSharedMemoryObjectNameLength = 32; // SHM_NAME_MAX on OSX ARM64, on other systems it's equal PATH_MAX (250)
+            // The POSIX shared memory object name must begin with '/'.  After that we just want something short (32) and unique.
+            Span<char> characters = stackalloc char[MaxSharedMemoryObjectNameLength];
+            char format = 'N';
+            bool success = Guid.NewGuid().TryFormat(characters, out int charsWritten, new ReadOnlySpan<char>(&format, 1));
+            Debug.Assert(success && charsWritten == MaxSharedMemoryObjectNameLength);
+            characters[0] = '/';
+            string mapName = new string(characters);
```
Comment (danmoseley) [inline src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs:184]: would it be helpful to prefix something so we know it's a .NET pipe? might be useful in debugging

eg instead of `/487a7997048425da6fc17a0f16f4bec` then `/dotnet97048425da6fc17a0f16f4bec` - still loads of entropy.

File: src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs [hunk 3]
```diff
@@ -170,11 +171,17 @@ private static FileStream CreateSharedBackingObject(Interop.Sys.MemoryMappedProt
...
+        private static unsafe FileStream? CreateSharedBackingObjectUsingMemory(
            Interop.Sys.MemoryMappedProtections protections, long capacity, HandleInheritability inheritability)
         {
-            // The POSIX shared memory object name must begin with '/'.  After that we just want something short and unique.
-            string mapName = string.Create(null, stackalloc char[128], $"/corefx_map_{Guid.NewGuid():N}");
+            const int MaxSharedMemoryObjectNameLength = 32; // SHM_NAME_MAX on OSX ARM64, on other systems it's equal PATH_MAX (250)
+            // The POSIX shared memory object name must begin with '/'.  After that we just want something short (32) and unique.
+            Span<char> characters = stackalloc char[MaxSharedMemoryObjectNameLength];
+            char format = 'N';
+            bool success = Guid.NewGuid().TryFormat(characters, out int charsWritten, new ReadOnlySpan<char>(&format, 1));
```
Comment (danmoseley) [inline src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs:181]: `bool success = Guid.NewGuid().TryFormat(characters, out int charsWritten, "N".AsSpan());` avoids unsafe code?

**Review by danmoseley — APPROVED** [review]

**Comment by danmoseley** [comment]
servicing candidate?

File: src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs (hunk 3 above)
Comment (adamsitnik) [reply src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs:181]: To be honest I was sure that it would allocate more memory, but seems not:

```cs
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System;

namespace MapName
{
    internal class Program
    {
        static unsafe void Main(string[] args) => BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args);
    }

    public class Strings
    {
        [Benchmark(Baseline = true)]
        public string Current() => string.Create(null, stackalloc char[128], $"/corefx_map_{Guid.NewGuid():N}");

        [Benchmark]
        public unsafe string Mutable()
        {
            string simple = Guid.NewGuid().ToString("N");
            fixed (char* first = simple)
            {
                first[0] = '/';
            }
            return simple;
        }

        [Benchmark]
        public unsafe string Fix()
        {
            Span<char> characters = stackalloc char[32];
            char format = 'N';
            Guid.NewGuid().TryFormat(characters, out int charsWritten, new ReadOnlySpan<char>(&format, 1));
            characters[0] = '/';
            return new string(characters);
        }

        [Benchmark]
        public string Dan()
        {
            Span<char> characters = stackalloc char[32];
            Guid.NewGuid().TryFormat(characters, out int charsWritten, "N".ToString());
            characters[0] = '/';
            return new string(characters);
        }
    }
}
```

|  Method |     Mean |    Error |  StdDev | Ratio |  Gen 0 | Allocated |
|-------- |---------:|---------:|--------:|------:|-------:|----------:|
| Current | 192.2 ns | 23.32 ns | 1.28 ns |  1.00 | 0.0134 |     112 B |
| Mutable | 157.2 ns | 16.23 ns | 0.89 ns |  0.82 | 0.0105 |      88 B |
|     Fix | 165.5 ns | 27.24 ns | 1.49 ns |  0.86 | 0.0105 |      88 B |
|     Dan | 167.4 ns | 17.69 ns | 0.97 ns |  0.87 | 0.0105 |      88 B |

File: src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs (hunk 2 above)
Comment (stephentoub) [inline src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs:184]: Why did you remove use of string.Create?

Comment (stephentoub) [reply src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs:184]: Ah, because Guid.NewGuid().ToString("N").Length == 32 and you need the slash first?

File: src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs (hunk 3 above)
Comment (stephentoub) [inline src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs:181]:
```suggestion
            bool success = Guid.NewGuid().TryFormat(characters, out int charsWritten, "N");
```

File: src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs:184]: In that case, you can format directly into the string, e.g.
```C#
string mapName = string.Create(MaxSharedMemoryObjectNameLength, 0, (span, state) =>
{
    Guid.NewGuid().TryFormat(span, out int charsWritten, "N");
    Debug.Assert(charsWritten == MaxSharedMemoryObjectNameLength);
    span[0] = '/';
});
```

File: src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs:181]: Keep in mind that stackalloc's have more cost the larger the amount of space unless you've specified SkipLocalsInit, which the benchmark above doesn't have but all of our core assemblies do.  That's why "Current" in your chart shows up as so much larger than the others (that, and it's generating a longer string than the rest because it still has the longer prefix).

```C#
using System;
using System.Runtime.CompilerServices;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

[module: SkipLocalsInit]

public class Program
{
    public static void Main(string[] args) => BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args);

    [Benchmark(Baseline = true)]
    public string Current() => string.Create(null, stackalloc char[128], $"/{Guid.NewGuid():N}"); // 33 char string, so not 100% comparable to the rest

    [Benchmark]
    public unsafe string Mutable()
    {
        string simple = Guid.NewGuid().ToString("N");
        fixed (char* first = simple)
        {
            first[0] = '/';
        }
        return simple;
    }

    [Benchmark]
    public unsafe string Fix()
    {
        Span<char> characters = stackalloc char[32];
        char format = 'N';
        Guid.NewGuid().TryFormat(characters, out int charsWritten, new ReadOnlySpan<char>(&format, 1));
        characters[0] = '/';
        return new string(characters);
    }

    [Benchmark]
    public string Dan()
    {
        Span<char> characters = stackalloc char[32];
        Guid.NewGuid().TryFormat(characters, out int charsWritten, "N");
        characters[0] = '/';
        return new string(characters);
    }

    [Benchmark]
    public string Steve()
    {
        return string.Create(32, 0, (span, state) =>
        {
            Guid.NewGuid().TryFormat(span, out int charsWritten, "N");
            span[0] = '/';
        });
    }
}
```

|  Method |     Mean |   Error |  StdDev | Ratio |
|-------- |---------:|--------:|--------:|------:|
| Current | 192.8 ns | 2.17 ns | 1.92 ns |  1.00 |
| Mutable | 186.9 ns | 1.12 ns | 0.94 ns |  0.97 |
|     Fix | 191.9 ns | 0.66 ns | 0.62 ns |  1.00 |
|     Dan | 190.7 ns | 1.50 ns | 1.33 ns |  0.99 |
|   Steve | 188.4 ns | 1.77 ns | 1.65 ns |  0.98 |

File: src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs (hunk 2 above)
Comment (jkotas) [reply src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs:184]: You can keep the entropy by using Base64 encoding. Then the name can be both helpful and have a lot of entropy, even more than it has today.

File: src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs]: (Though the cost of this operation is almost certainly dwarfed entirely by everything else that needs to happen as part of creating the MMF.)

Comment (adamsitnik) [reply src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs]: This conversation reminds me of https://meta.wikimedia.org/wiki/Cunningham%27s_Law 

> the best way to get the right answer on the internet is not to ask a question; it's to post the wrong answer.

File: src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs [hunk 4]
```diff
@@ -170,11 +171,18 @@ private static FileStream CreateSharedBackingObject(Interop.Sys.MemoryMappedProt
...
-            // The POSIX shared memory object name must begin with '/'.  After that we just want something short and unique.
-            string mapName = string.Create(null, stackalloc char[128], $"/corefx_map_{Guid.NewGuid():N}");
+            const int MaxSharedMemoryObjectNameLength = 32; // SHM_NAME_MAX on OSX ARM64, on other systems it's equal PATH_MAX (250)
+            // The POSIX shared memory object name must begin with '/'.  After that we just want something short (32) and unique.
+            string mapName = string.Create(MaxSharedMemoryObjectNameLength, 0, (span, state) =>
+            {
+                Guid.NewGuid().TryFormat(span, out int charsWritten, "N");
+                Debug.Assert(charsWritten == MaxSharedMemoryObjectNameLength);
+                span[0] = '/';
+                "dotnet".CopyTo(span.Slice(1));
```
Comment (stephentoub) [inline src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs:184]:
```suggestion
                "/dotnet".CopyTo(span);
```

File: src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs (hunk 2 above)
Comment (danmoseley) [reply src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs]: Great idea I didn't think of.

File: src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs (hunk 4 above)
Comment (adamsitnik) [reply src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs]: @jkotas do you mean getting the 16 bytes out of Guid and mapping it into a string using the base64 encoding?

I'll check on Monday whether all characters used by the base64 encoding are valid for map names

Comment (jkotas) [reply src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs]: You can use proper random number generator `RandomNumberGenerator.Fill` if you go with Base64 encoding.

Note that Guid contains less entropy than what you may think. From https://docs.microsoft.com/en-us/dotnet/api/system.guid.newguid?view=net-6.0 : The generated GUID contains 122 bits of strong entropy.

Comment (adamsitnik) [reply src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs]: @jkotas I've given your idea a try. It works, but:

* using `RandomNumberGenerator.Fill`, requires to add a new dependency to `System.Security.Cryptography`. Since I have to backport this fix to 6.0 I am hesitant of adding any new dependencies.
* I've [tested](https://github.com/dotnet/runtime/pull/64099/commits/64bb94b4ce1f37490940f245dfcda86b2c7f2683) and using an additional `/` in the name (produced by Base64) fails with EINVAL. Of course we can easily replace it.

I would prefer to keep using `Guid`, but also implement a simply retry logic for `EEXIST` (the error we get when a map with such name already exists). @jkotas what do you think?


<details>

```cs
const int MaxSharedMemoryObjectNameLength = 32; // SHM_NAME_MAX on OSX ARM64, on other systems it's equal PATH_MAX (250)
// The POSIX shared memory object name must begin with '/'.  After that we just want something short (32) and unique.
string mapName = string.Create(MaxSharedMemoryObjectNameLength, 0, (span, state) =>
{
    const string Prefix = "/dotnet_map_";
    Prefix.CopyTo(span);
    const int ByteCount = 15;
    Debug.Assert(ByteCount == (MaxSharedMemoryObjectNameLength - Prefix.Length) / 4 * 3); // Base64 turns every 3 bytes into 4 characters
    Span<byte> buffer = stackalloc byte[ByteCount];
    RandomNumberGenerator.Fill(buffer);
    Convert.TryToBase64Chars(buffer, span.Slice(Prefix.Length), out int charsWritten);
    Debug.Assert(charsWritten == MaxSharedMemoryObjectNameLength - Prefix.Length);

    for (int i = Prefix.Length; i < span.Length; i++)
    {
        if (span[i] == '/' || span[i] == '+') // used by Base64 encoding, but invalid for shm_open (EINVAL)
        {
            span[i] = '_';
        }
    }
});
```

</details>

Comment (jkotas) [reply src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs]: Ok, keeping Guid sounds good to me.

File: src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs [hunk 5]
```diff
@@ -191,22 +189,31 @@ private static FileStream CreateSharedBackingObject(Interop.Sys.MemoryMappedProt
...
+
+                if (fd.IsInvalid)
                 {
-                    // If ShmOpen is not supported, fall back to file backing object.
-                    // Note that the System.Native shim will force this failure on platforms where
-                    // the result of native shm_open does not work well with our subsequent call
-                    // to mmap.
-                    return null;
-                }
+                    Interop.ErrorInfo errorInfo = Interop.Sys.GetLastErrorInfo();
```
Comment (adamsitnik) [inline src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs:202]:
```suggestion
                    Interop.ErrorInfo errorInfo = Interop.Sys.GetLastErrorInfo();
                    fd.Dispose();
```
